Define BaseExperiment._transpiled_circuits and deprecate BaseExperiment._postprocess_transpiled_circuits

kevinsung opened the pull request:
Fixes #669.

<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Adds `BaseExperiment._transpiled_circuits` to enable customization of transpilation. This makes `BaseExperiment._postprocess_transpiled_circuits` obsolete so that is deprecated with a version 0.4.0 deadline.


### Details and comments
@chriseclectic I didn't find any issues with composite experiments, but let me know if there are any specific tests I should add.

nkanazawa1989 commented on qiskit_experiments/framework/base_experiment.py:
```diff
@@ -269,8 +269,7 @@ def run(
         transpile_opts["initial_layout"] = list(experiment.physical_qubits)
-        circuits = transpile(experiment.circuits(), experiment.backend, **transpile_opts)
-        experiment._postprocess_transpiled_circuits(circuits, **run_options)
+        circuits = self._transpile(experiment.circuits(), experiment.backend, **transpile_opts)
```
Could you please rename this to `transpiled_circuit` for readability?

kevinsung replied:
Do you mean `transpiled_circuits`? I think that name is confusing because it makes it seem like this is a version of `BaseExperiment.circuits` which just generates the transpiled versions instead. In reality, it takes the circuits as input and has the same function signature as `qiskit.transpile`. The analogy with `qiskit.transpile` is the reason I think the current name is the best. What do you think?

nkanazawa1989 replied:
Indeed this is nitpick and not exposed to external users. However for developers logical vs physical circuit is really important. If a developer reads this line
```python
jobs = experiment._run_jobs(circuits, **run_opts)
```
one may wonder `circuits` is physical or not (of course it is physical).

nkanazawa1989 commented on qiskit_experiments/framework/base_experiment.py:
```diff
@@ -479,9 +478,27 @@ def set_analysis_options(self, **fields):
-        """Additional post-processing of transpiled circuits before running on backend"""
-        pass
+    def _transpile(
+        self, circuits: List[QuantumCircuit], backend: Backend, **transpile_options
```
Probably backend is not necessary since you can access to it via `self.backend` (also this will bypass `_set_backend` hook). Transpile options and circuit too. Likely you want to make it static method -- is there any plan to hack this method? Something like injecting circuit which is not generated by this experiment instance?

kevinsung replied:
It is necessary because `self.backend` could be None, in which case the backend is passed to `BaseExperiment.run` and forwarded to this function.

kevinsung replied:
It should not be a static method because transpilation can depend on the experiment instance.

nkanazawa1989 replied:
This is not correct because existence of backend is guaranteed by these lines. Seems like this allows a hack that injecting backend which is not associated with the instance. Currently you are using `self` to only access deprecated method.
https://github.com/Qiskit/qiskit-experiments/blob/317b964b8cbde0febcda8696fd4bd5cce3490e0a/qiskit_experiments/framework/base_experiment.py#L251-L259

nkanazawa1989 replied:
Perhaps you are assuming the situation the user directly call this method of the instance initialized without backend to check the physical circuit? Then I agree it's worth having backend here. But I think circuit and transpile options can come from self.

kevinsung replied:
Still, the backend needs to be forwarded because it uses `experiment.backend` which could be different from `self.backend`.

kevinsung replied:
> Perhaps you are assuming the situation the user directly call this method of the instance initialized without backend to check the physical circuit? Then I agree it's worth having backend here. But I think circuit and transpile options can come from self.

Yes, exactly. If the user calls `run` with a backend specified, then using `self.backend` would be incorrect. I actually this this a bit confusing, and that only way of passing the backend should be supported. That would also eliminate the need for the complicated backend selection code. But that is a separate issue.

nkanazawa1989 commented on qiskit_experiments/framework/base_experiment.py:
```diff
@@ -479,9 +478,27 @@ def set_analysis_options(self, **fields):
+        transpiled = transpile(circuits, backend, **transpile_options)
+
+        # TODO remove this deprecation after 0.3.0 release
+        if hasattr(self, "_postprocess_transpiled_circuits"):
```
Good catch. This is for custom experiment code that has already built by the community?

kevinsung replied:
Yes.

nkanazawa1989 reviewed (changes requested):
Code looks good to me but this is indeed conflicting with #667 @yaelbh could you please decide if merge this first or close this and make @kevinsung co-author of your PR.

yaelbh commented:
I suggest that the PR that's ready first will be merged first, and the other one will resolve the conflicts

kevinsung commented:
@chriseclectic I have eliminated the transpilation keyword arguments. As I explained above, `backend` is needed because for the case the user called `run` with a backend specified. I think `circuits` should remain as a keyword argument in order to have a clean separation of functionality with `BaseExperiment.circuits`.

yaelbh commented on qiskit_experiments/framework/base_experiment.py:
```diff
@@ -267,13 +267,10 @@ def run(
-        transpile_opts["initial_layout"] = list(experiment.physical_qubits)
-        circuits = transpile(experiment.circuits(), experiment.backend, **transpile_opts)
-        experiment._postprocess_transpiled_circuits(circuits, **run_options)
+        transpiled_circuits = self._transpile(experiment.circuits(), experiment.backend)
```
I find it confusing that `self` and `experiment` exist and are being used side-by-side. In the current code, from the moment that we define `experiment`, we forget about `self` and stick with `experiment` until the end of the method. I'm not fond of this situation, because it still calls for bugs related to someone not noticing this fact and using `self` where it should be `experiment` or vice versa, but I can live with it. But if they're really both used, then it's even more confusing. So, if you can replace `self` here with `experiment`, then please do. Otherwise, it's important to document exactly why the mixture is necessary. And the best is, of course, if there is a way to get rid of `experiment` in the first place. But I don't see such a way. Is it that important to support setting of backend and analysis in run time?

yaelbh commented on qiskit_experiments/framework/base_experiment.py:
```diff
@@ -479,9 +476,27 @@ def set_analysis_options(self, **fields):
+        transpile_opts["initial_layout"] = list(self.physical_qubits)
+        transpiled = transpile(circuits, backend, **transpile_opts)
+
+        # TODO remove this deprecation after 0.3.0 release
```
Do we have to go through the deprecation process for private methods? (`_postprocess_transpiled_circuits` starts with an underscore).

yaelbh reviewed (commented):
Thanks @kevinsung 😃.
I intentionally mark "Comment" and not "Request changes", to make the approvals of Chris and Naoki sufficient for merging the PR.

yaelbh commented:
Please don't proceed with approving and merging this PR before we resolve https://github.com/Qiskit/qiskit-experiments/pull/667#issuecomment-1046178725.

kevinsung commented:
CI failure appears unrelated again, see https://github.com/Qiskit/qiskit-experiments/issues/682

chriseclectic commented on qiskit_experiments/framework/base_experiment.py:
```diff
@@ -362,6 +359,27 @@ def circuits(self) -> List[QuantumCircuit]:
         # values for any explicit experiment options that affect circuit
         # generation
 
+    def _transpiled_circuits(self, backend: Backend) -> List[QuantumCircuit]:
```
```suggestion
    def _transpiled_circuits(self) -> List[QuantumCircuit]:
```
The backend should come from `self` as in @yaelbh's PR

kevinsung replied:
Done.

chriseclectic commented on qiskit_experiments/framework/base_experiment.py:
```diff
@@ -362,6 +359,27 @@ def circuits(self) -> List[QuantumCircuit]:
+        """
+        transpile_opts = copy.copy(self.transpile_options.__dict__)
+        transpile_opts["initial_layout"] = list(self.physical_qubits)
+        transpiled = transpile(self.circuits(), backend, **transpile_opts)
```
```suggestion
        transpiled = transpile(self.circuits(), self.backend, **transpile_opts)
```

kevinsung replied:
Done.

chriseclectic commented on releasenotes/notes/transpile-617bd3a4e6f1c0d8.yaml:
```diff
@@ -0,0 +1,10 @@
+---
+features:
+  - |
+    Adds `BaseExperiment._transpile` which is called to transpile the
```
This comment needs updating to refer to `_transpiled_circuits` method and slightly changed functionality (it returns a list of the transpiled circuits to be run on a backend).
```suggestion
    Adds :meth:`.BaseExperiment._transpiled_circuits` which is called to transpile the
```

kevinsung replied:
Done.

chriseclectic commented on releasenotes/notes/transpile-617bd3a4e6f1c0d8.yaml:
```diff
@@ -0,0 +1,10 @@
+deprecations:
+  - |
+    `BaseExperiment._postprocess_transpiled_circuits` is deprecated and will
+    be removed in the 0.4.0 release. Use `BaseExperiment._transpile` instead.
```
This comment also needs updating to refer to the correct method. You should use RST links as well eg
```
:meth:`.BaseExperiment._transpiled_circuits`
```

kevinsung replied:
Done.

chriseclectic reviewed (commented):
This looks very close. I would suggest merging #667 first, then rebasing this again after.

kevinsung commented:
@chriseclectic I've addressed your comments. It might make sense to merge this before #667 because #667 actually includes this change and does more on top.

chriseclectic commented on releasenotes/notes/transpile-617bd3a4e6f1c0d8.yaml:
```diff
@@ -0,0 +1,11 @@
+---
+features:
+  - |
+    Adds :meth:`BaseExperiment._transpiled_circuits` which returns a list of
```
```suggestion
    Adds :meth:`.BaseExperiment._transpiled_circuits` which returns a list of
```

chriseclectic commented on releasenotes/notes/transpile-617bd3a4e6f1c0d8.yaml:
```diff
@@ -0,0 +1,11 @@
+    transpilation.
+deprecations:
+  - |
+    :meth:`BaseExperiment._postprocess_transpiled_circuits` is deprecated and
```
```suggestion
    ``BaseExperiment._postprocess_transpiled_circuits`` is deprecated and
```

chriseclectic commented on releasenotes/notes/transpile-617bd3a4e6f1c0d8.yaml:
```diff
@@ -0,0 +1,11 @@
+  - |
+    :meth:`BaseExperiment._postprocess_transpiled_circuits` is deprecated and
+    will be removed in the 0.4.0 release.
+    Use :meth:`BaseExperiment._transpiled_circuits` instead.
```
```suggestion
    Use :meth:`.BaseExperiment._transpiled_circuits` instead.
```

chriseclectic reviewed (changes requested):
Minor formatting

chriseclectic reviewed (approved):
@kevinsung sounds good, we can merge this one first
